domain/src/test/java/org/oppia/domain/audio/AudioPlayerControllerTest.kt

```diff
@@ -0,0 +1,252 @@
+import android.content.Context
+import android.net.Uri
+import androidx.lifecycle.Observer
+import androidx.test.core.app.ApplicationProvider
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import dagger.BindsInstance
+import dagger.Component
+import dagger.Module
+import dagger.Provides
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.test.TestCoroutineDispatcher
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.Mock
+import org.mockito.junit.MockitoJUnit
+import org.mockito.junit.MockitoRule
+import org.mockito.Mockito.verify
+import org.oppia.util.logging.EnableConsoleLog
+import org.oppia.util.logging.EnableFileLog
+import org.oppia.util.logging.GlobalLogLevel
+import org.oppia.util.logging.LogLevel
+import org.oppia.util.threading.BackgroundDispatcher
+import org.oppia.util.threading.BlockingDispatcher
+import org.robolectric.Shadows
+import org.robolectric.annotation.Config
+import javax.inject.Inject
+import javax.inject.Singleton
+import com.google.common.truth.Truth.assertThat
+import org.mockito.ArgumentCaptor
+import org.mockito.Captor
+import org.mockito.Mockito.atLeastOnce
+import org.oppia.util.data.AsyncResult
+import org.robolectric.shadows.ShadowMediaPlayer
+import org.robolectric.shadows.util.DataSource
+import javax.inject.Qualifier
+
+/** Tests for [AudioPlayerControllerTest]. */
+@RunWith(AndroidJUnit4::class)
+@Config(manifest = Config.NONE)
+class AudioPlayerControllerTest {
+
+  @Rule
+  @JvmField
+  val mockitoRule: MockitoRule = MockitoJUnit.rule()
+
+  @Mock
+  lateinit var mockAudioPlayerObserver: Observer<AsyncResult<AudioPlayerController.PlayProgress>>
+
+  @Captor
+  lateinit var audioPlayerResultCaptor: ArgumentCaptor<AsyncResult<AudioPlayerController.PlayProgress>>
+
+  @Inject
+  lateinit var context: Context
+
+  @Inject
+  lateinit var audioPlayerController: AudioPlayerController
+
+  private lateinit var shadowMediaPlayer: ShadowMediaPlayer
+
+  private val TEST_URL = "https://www.soundhelix.com/examples/mp3/SoundHelix-Song-1.mp3"
+
+  @Before
+  fun setup() {
+    setUpTestApplicationComponent()
+    addMediaInfo()
+    shadowMediaPlayer = Shadows.shadowOf(audioPlayerController.getTestMediaPlayer())
+    shadowMediaPlayer.dataSource = DataSource.toDataSource(context , Uri.parse(TEST_URL))
+  }
+
+
+  @Test
+  fun testAudioPlayer_successfulInitialize_reportsSuccessfulInit() {
+    audioPlayerController.initializeMediaPlayer(TEST_URL)
+
+    shadowMediaPlayer.invokePreparedListener()
+
+    assertThat(shadowMediaPlayer.isPrepared).isTrue()
+    assertThat(audioPlayerController.isPlaying()).isFalse()
+  }
+
+  @Test
+  fun testAudioPlayer_play_isPlaying() {
+    arrangeMediaPlayer()
+
+    audioPlayerController.play()
+
+    assertThat(audioPlayerController.isPlaying()).isTrue()
+    assertThat(shadowMediaPlayer.isReallyPlaying).isTrue()
```

Comment:
Haha I like how this tests reads:
- Is my object telling me it's working? Yes.
- But is it REALLY working? \**checks the system.*\* Still yes.